### ISSUE TITLE
Build Fails on “reading sources...” Without Error Message

### ISSUE THREAD

**Issue by gbroques** [readthedocs/readthedocs.org#8723]
## Details

* Read the Docs project URL: https://readthedocs.org/projects/openafpm-cad-core/
* Build URL (if applicable): https://readthedocs.org/projects/openafpm-cad-core/builds/15393876/
* Read the Docs username (if applicable): gbroques

For some reason my build fails on "reading sources... ":
```
[rtd-command-info] start-time: 2021-11-28T21:11:30.196140Z, end-time: 2021-11-28T21:11:31.576485Z, duration: 1, exit-code: 1
/home/docs/checkouts/readthedocs.org/user_builds/openafpm-cad-core/conda/latest/bin/python -m sphinx -T -E -b html -d _build/doctrees -D language=en . _build/html
Running Sphinx v4.1.1
loading translations [en]... done
making output directory... done
[autosummary] generating autosummary for: index.rst
Creating file openafpm_cad_core/openafpm_cad_core.rst.
Creating file openafpm_cad_core/openafpm_cad_core.create_spreadsheet_document.rst.
Creating file openafpm_cad_core/openafpm_cad_core.gui.rst.
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 1 source files that are out of date
updating environment: [new config] 4 added, 0 changed, 0 removed
reading sources... [ 25%] index
reading sources... [ 50%] openafpm_cad_core/openafpm_cad_core
```

I'm unable to reproduce this error locally, and I don't see any error message -- making this hard to debug.

It seems like someone else had a similar issue and posted on Stack Overflow about it:
https://stackoverflow.com/questions/42470735/readthedocs-build-fails-on-reading-sources-without-error

For them, the issue seemed to resolve itself all of the sudden without doing anything to fix it.

Any help on troubleshooting this is greatly appreciated! :-)

**Comment by humitos** [comment]
I took a look at your latest builds and I didn't find anything useful in the logs :/ . I see where it is failing but I'm not sure why

```
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build readthedocs.doc_builder.environments:587[2189]: WARNING (Build) [openafpm-cad-core:latest] Command /home/docs/checkouts/readthedocs.org/user_builds/
openafpm-cad-core/conda/latest/bin/python -m sphinx -T -E -b html -d _build/doctrees -D language=en . _build/html failed:
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build Running Sphinx v4.1.1
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build loading translations [en]... done
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build making output directory... done
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build [autosummary] generating autosummary for: index.rst
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build Creating file openafpm_cad_core/openafpm_cad_core.rst.
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build Creating file openafpm_cad_core/openafpm_cad_core.create_spreadsheet_document.rst.
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build Creating file openafpm_cad_core/openafpm_cad_core.gui.rst.
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build building [mo]: targets for 0 po files that are out of date
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build building [html]: targets for 1 source files that are out of date
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build updating environment: [new config] 4 added, 0 changed, 0 removed
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build reading sources... [ 25%] index
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build reading sources... [ 50%] openafpm_cad_core/openafpm_cad_core
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build Traceback (most recent call last):
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build   File "/home/docs/checkouts/readthedocs.org/readthedocs/projects/tasks.py", line 799, in run_build
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build     outcomes = self.build_docs()
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build   File "/home/docs/checkouts/readthedocs.org/readthedocs/projects/tasks.py", line 1250, in build_docs
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build     outcomes['html'] = self.build_docs_html()
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build   File "/home/docs/checkouts/readthedocs.org/readthedocs/projects/tasks.py", line 1268, in build_docs_html
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build     success = html_builder.build()
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build   File "/home/docs/checkouts/readthedocs.org/readthedocs/doc_builder/backends/sphinx.py", line 282, in build
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build     cmd_ret = self.run(
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build   File "/home/docs/checkouts/readthedocs.org/readthedocs/doc_builder/base.py", line 144, in run
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build     return self.build_env.run(*args, **kwargs)
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build   File "/home/docs/checkouts/readthedocs.org/readthedocs/doc_builder/environments.py", line 613, in run
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build     return super().run(*cmd, **kwargs)
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build   File "/home/docs/checkouts/readthedocs.org/readthedocs/doc_builder/environments.py", line 395, in run
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build     return self.run_command_class(cls=self.command_class, cmd=cmd, **kwargs)
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build   File "/home/docs/checkouts/readthedocs.org/readthedocs/doc_builder/environments.py", line 619, in run_command_class
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build     return super().run_command_class(*cmd, **kwargs)
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build   File "/home/docs/checkouts/readthedocs.org/readthedocs/doc_builder/environments.py", line 467, in run_command_class
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build     raise BuildEnvironmentWarning(msg)
Nov 29 02:42:53 build-large-i-0880d11cd9647764f /supervisord: build readthedocs.doc_builder.exceptions.BuildEnvironmentWarning: Command /home/docs/checkouts/readthedocs.org/user_builds/openafpm-cad-core/conda/latest/bin/python -m sphinx -T -E -b html -d _build/doctrees -D language=en . _build/html failed:
```

**Comment by gbroques** [comment]
> I took a look at your latest builds and I didn't find anything useful in the logs :/ . I see where it is failing but I'm not sure why

Thanks for taking a look!

I suppose I need to:

* Try harder to reproduce this error locally.
* Backtrack to when the build was working, and pinpoint what change caused this to start failing.

**Comment by astrojuanlu** [comment]
Hi @gbroques, I tried to have a quick look at this problem but I'm not even able to build the documentation locally using the Read the Docs images. You can do that with:

```
$ docker run --rm -ti --user 0 readthedocs/build:7.0 bash
```

However, the doc fails almost immediately:

```
>>> import FreeCAD
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'FreeCAD'
```

even though I can clearly see `FreeCAD.so` here:

```
/home/docs/.conda/envs/latest/lib/FreeCAD.so
```

maybe I failed at some basic intermediate step, but even though I have lots of experience with conda-forge packages, I don't know about the specifics of FreeCAD. If you give us instructions to try to build your docs locally, maybe we can help.

**Comment by gbroques** [comment]
> Hi @gbroques, I tried to have a quick look at this problem but I'm not even able to build the documentation locally using the Read the Docs images. You can do that with:
> 
> ```
> $ docker run --rm -ti --user 0 readthedocs/build:7.0 bash
> ```
> 
> However, the doc fails almost immediately:
> 
> ```
> >>> import FreeCAD
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> ModuleNotFoundError: No module named 'FreeCAD'
> ```
> 
> even though I can clearly see `FreeCAD.so` here:
> 
> ```
> /home/docs/.conda/envs/latest/lib/FreeCAD.so
> ```
> 
> maybe I failed at some basic intermediate step, but even though I have lots of experience with conda-forge packages, I don't know about the specifics of FreeCAD. If you give us instructions to try to build your docs locally, maybe we can help.

Hi @astrojuanlu, thanks for digging into this.

That's a specific quirk of using FreeCAD as a Python module.

There's a [add_conda_path_to_python_path.sh](https://github.com/gbroques/openafpm-cad-core/blob/master/add_conda_path_to_python_path.sh) that I run once locally to setup my conda environment.

Basically, you have to add the location of that conda `lib` directory to Python's `sys.path` or `PYTHONPATH` so it knows how to find `FreeCAD.so`.

I have special code for this in my [conf.py](https://github.com/gbroques/openafpm-cad-core/blob/master/docs/conf.py#L21-L27) for this to work on the Read the Docs server.

**Comment by gbroques** [comment]
We may be in the clear.

I have a [passing build now](https://readthedocs.org/projects/openafpm-cad-core/builds/15411533/)!

I was able to reproduce some strange error manifesting as a warning locally.

To fix it, I had to upgrade a conda package (`freecad==0.19.2` was bumped to `0.19.3`).

Then for some strange reason, I had to disable the Latex / PDF format since [that was failing](https://readthedocs.org/projects/openafpm-cad-core/builds/15411483/).

Additionally, I wiped the build environment.

I don't want to take up any more of your time!

Thanks for the help, I will re-open if needed! :)
